Track unmappable dims correctly in compute at.

csarofeen opened the pull request:
Fixes #1305

See comments in issue for more details.

csarofeen committed:
Track unmappable dims correctly in compute at.

naoyam reviewed (approved):
The fix makes sense, but looks like the real problem exists in `ComputeAtRootDomainMap`. I don't understand why `t3` and `t4` are completely inlined. I'll look into it, but feel free to merge this PR if this is urgent.

naoyam commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -19604,6 +19604,31 @@ TEST(NVFuserTest, FusionNonDivisibleSplitVectorize2_CUDA) {
   testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
 }
 
+TEST(NVFuserTest, Issue1305Repro_CUDA) {
```
nit: `TEST_F(NVFuserTest, FusionIssue1305Repro_CUDA) {`.

I wonder we really need the `Fusion` suffix.

csarofeen replied:
Probably not but best to be consistent

naoyam commented:
In the repro, the edge from `T2` to `T4` makes `ComputeAtRootMapBuilder` incorrectly map a pair of `T2` and `T4` and another pair of `T2` and `T3`, resulting in mapping between `T4` and `T3`. This should be fixable.

naoyam commented:
#1327 fixes the problem with `ComputeAtRootMapBuilder`. I confirmed the repro works fine without the fix in this PR.

csarofeen commented:
At a minimum this should improve performance. There's bad performance scaling with the number of expressions which does seem to be the compute at pass. So I think this PR can't hurt.

csarofeen committed:
Fix test names.

naoyam committed:
Fix ComputeAtRootDomainMap (#1327)

Once a pair of domains is determined to be invalid to map, keep that
information during the traversal in ComputeAtRootDomainMapBuilder. This
is to avoid indirectly cause invalid mappings. See issue #1305 for an
example.

csarofeen committed:
Merge branch 'devel' into issue_1305
